pkg/kubelet/resources.go

```diff
@@ -0,0 +1,57 @@
+/*
+Copyright 2016 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package kubelet
+
+import (
+	"fmt"
+
+	"github.com/golang/glog"
+
+	"k8s.io/kubernetes/pkg/api"
+)
+
+func (kl *Kubelet) defaultPodLimitsForDownwardApi(pod *api.Pod) (*api.Pod, error) {
+	capacity := make(api.ResourceList)
+	lastUpdatedNodeObject := kl.lastUpdatedNodeObject.Load()
+	if lastUpdatedNodeObject == nil {
+		glog.Errorf("Failed to find node object in cache. Expected a non-nil object in the cache.")
+	} else {
+		capacity = lastUpdatedNodeObject.(*api.Node).Status.Capacity
+	}
+	podCopy, err := api.Scheme.Copy(pod)
+	if err != nil {
+		return nil, fmt.Errorf("failed to perform a deep copy of pod object. Error: %v", err)
+	}
+	pod = podCopy.(*api.Pod)
+	for idx, c := range pod.Spec.Containers {
+		// If limits are not set at all, set them to equal capacity.
+		for resource, cap := range capacity {
+			if _, exists := c.Resources.Limits[resource]; !exists {
+				pod.Spec.Containers[idx].Resources.Limits[resource] = cap
+			}
+		}
+		// The the existing limits are <=0, update them to equal capacity.
+		for resource, quantity := range c.Resources.Limits {
+			if quantity.MilliValue() <= 0 {
```

Comment:
It does not seem to be an issue for cpu or memory, but could be for storage. 
Is it (im)practical to assume that kube nodes might have petabytes storage capacity? if practical, In that case, it might be better to check for overflow for storage. I am saying this because the copy generated here is being passed to all volumes not just downward api volumes, as I understand. or may be just limit the resource check for cpu and memory to avoid any issues with storage.